[mhaikelll] iP

mhaikelll opened the pull request:
# Huhuhuharis

> “Your mind is for having ideas, not holding them.” – David Allen

Huhuhuharis frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use 

All you need to do is,

1. download it from [here](https://github.com/mhaikelll/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😗 

And it is **FREE!**

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```
public static void main(String[] args) throws HuhuhuharisException {
        new Huhuhuharis("./data/huhuhuharis.txt").run();
}

public void run() {
        ui.showWelcomeMessage();
        while (true) {
            String input = ui.getUserInput();
            String reply = chatResponse(input);
            ui.showTaskList(reply);
            if (input.equals("bye")) {
                break;
            }
        }
        Storage.saveTasks(taskList.getTasks());
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

mhaikelll committed:
Renamed file, greet, exit

mhaikelll committed:
Renamed file

mhaikelll committed:
Echo Response

mhaikelll committed:
Add, List added

mhaikelll committed:
Add mark as done logic

mhaikelll committed:
Add todo, event and deadline classes

mhaikelll committed:
Add error handling

mhaikelll committed:
Add support for deleting tasks

mhaikelll committed:
Add automated text UI testing

mhaikelll committed:
Add storage class, save and load tasks

mhaikelll committed:
Merge branch-Level-7 into master

mhaikelll committed:
Add parser class, handle dates and times

mhaikelll committed:
Merge branch-Level-8 into master

mhaikelll committed:
Add more OOP classes.

mhaikelll committed:
Add package Huhuhuharis

mhaikelll committed:
Merge add gradle support to master

mhaikelll committed:
Add gradle support

WhiteBear82 commented on src/main/java/Huhuhuharis/Huhuhuharis.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    public static String deleteTask(String input) {
+        int taskId = Integer.parseInt(input.split(" ")[1]) - 1;
```
It looks like there isn’t a validation check for taskId before accessing taskList. If the input isn’t a valid number or is out of bounds, it might cause the program to crash. It might be helpful to add a check to ensure taskId is within a valid range before attempting to remove the task. That way, users won’t run into unexpected errors.

WhiteBear82 commented on src/main/java/Huhuhuharis/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+        if (tasks == null) {
+            throw new NullPointerException("List cannot be null.");
+        }
+        tasks.addAll(taskLists);
```
Correct me if I'm wrong, but the logic here seems a bit reversed. Perhaps taskLists.addAll(tasks) would work better to ensure tasks loaded from storage are added correctly to the existing list?

WhiteBear82 commented on src/main/java/Huhuhuharis/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+
+public class TaskList {
+
+    private ArrayList<Task> taskLists;
```
The variable name taskLists seems a bit misleading since it holds a single list of tasks, not multiple lists. Would it be clearer to rename it to tasks? Just a suggestion for clarity!

WhiteBear82 commented on src/main/java/Huhuhuharis/Storage.java:
```diff
@@ -0,0 +1,73 @@
+
+public class Storage {
+
+    private String path;
```
It seems this variable is never used in the class. If it's unnecessary, would it be better to remove it for cleaner code?

WhiteBear82 commented on src/main/java/Huhuhuharis/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+        return listCount;
+    }
+
+    public String fullList() {
```
Just a suggestion, but using StringBuilder instead of string concatenation inside a loop could improve performance. It might be worth considering for efficiency!

```java
StringBuilder fullList = new StringBuilder("Here are the tasks in your list:\n");
for (int i = 0; i < listCount; i++) {
    fullList.append(i + 1).append(".").append(taskLists.get(i)).append("\n");
}
return fullList.toString();
```

WhiteBear82 commented on src/main/java/Huhuhuharis/Huhuhuharis.java:
```diff
@@ -0,0 +1,119 @@
+    public Huhuhuharis(String path) throws HuhuhuharisException {
+        ui = new Ui();
+        storage = new Storage("./data/huhuhuharis.txt");
+        taskList = new TaskList(Storage.loadTasks());
```
It looks like tasks are being loaded twice. Would it be possible to load the tasks once to avoid duplication? Something like:
```java
taskList = new TaskList(Storage.loadTasks());
```

WhiteBear82 commented on src/main/java/Huhuhuharis/Storage.java:
```diff
@@ -0,0 +1,73 @@
+                writer.write(task.saveToFile() + "\n");
+            }
+            writer.close();
+        } catch (IOException e) {
```
It might be helpful to catch more specific exceptions in addition to IOException and provide additional context in the error messages. For example, you could separate FileNotFoundException and IOException and give a bit more detail on the error. It could look something like this:
```java
try {
    // file operations
} catch (FileNotFoundException e) {
    System.out.println("Error: The specified file was not found.");
} catch (IOException e) {
    System.out.println("Error: An unexpected I/O error occurred.");
}
```

WhiteBear82 reviewed (commented):
Your attention to detail is evident throughout the project. For example, the handling of specific task types with their own classes not only keeps the code clean but also makes it scalable for future features. The exception handling is robust, and I noticed that you're mindful of giving clear error messages, which enhances debugging and user experience. You've also ensured that the UI is welcoming and intuitive, which contributes positively to the overall user experience by making interactions feel smooth and straightforward. Keep up the great work!

mhaikelll committed:
Add JUnit tests

mhaikelll committed:
Add jar file

mhaikelll committed:
Comply with standard

mhaikelll committed:
Merge coding standard

mhaikelll committed:
Add find logic

mhaikelll committed:
Merge level 9 with master

Jasonzhou97 commented on src/main/java/Huhuhuharis/Parser.java:
```diff
@@ -0,0 +1,29 @@
+    public static LocalDateTime strToDateTime(String str) throws DateTimeParseException {
+        try {
+            str = str.replaceAll("am", "AM").replaceAll("pm", "PM");
+            System.out.println("Trying to parse date-time: " + str);
```
Looks good, maybe can reflect in GUI instead of printing?

Jasonzhou97 commented on src/main/java/Huhuhuharis/Storage.java:
```diff
@@ -0,0 +1,74 @@
+                String line = sc.nextLine();
+                String[] parts = line.split(" \\| ");
+                Task task = null;
+                switch (parts[0]) {
```
I think switch and case should be at same indentation level?

Jasonzhou97 commented on src/main/java/Huhuhuharis/Huhuhuharis.java:
```diff
@@ -0,0 +1,139 @@
+        return result.toString();
+    }
+
+    public static String markTask(String input) {
```
Perhaps can add an exception handling here that handles if input length is less than 2?

Jasonzhou97 reviewed (commented):
Looks good overall!

mhaikelll committed:
Add Javadoc comments

mhaikelll committed:
Merge add javadoc to master

mhaikelll committed:
Add GUI to chatbot

mhaikelll committed:
Merge level 10 to master

mhaikelll committed:
Add priority handling logic with levels

Without priority levels, users cannot distinguish the importance
of varying tasks, making it harder to focus on urgent tasks.

Provide a method to assign priority levels (High, Medium, Low)
to tasks with a handlePriority(String input) function to parse
and apply the command. Save the changes to storage to ensure
persistence.

This keeps the implementation simple with the existing task
mmanagement structure which follows a structured command style
to minimize user errors.

mhaikelll committed:
Add assertions to ensure input validation and prevent crashes

Without asserions, potential crashes may occur due to invalid inputs,
affecting critical methods within the app.

Added assertions to verify non-null inputs, correct parsing of inputs,
checking of task indices and correct parsing of task numbers.

This improves code robustness by catching issues early, helping to detect
logic errors during development, preventing the risk of runtime exceptions.

mhaikelll committed:
Merge pull request #1 from mhaikelll/branch-A-Assertions

Add assertions to ensure input validation and prevent crashes

mhaikelll committed:
Improve code quality through renaming packages and editing Javadocs

mhaikelll committed:
Merge pull request #2 from mhaikelll/branch-A-CodeQuality

Improve code quality through renaming packages and editing Javadocs

mhaikelll committed:
Finalize features to the chatbot for release

mhaikelll committed:
Update README.md

mhaikelll committed:
Update README.md v2

mhaikelll committed:
Update README.md v3

mhaikelll committed:
Minor changes to main window and docs

mhaikelll committed:
Merge branch 'master' of https://github.com/mhaikelll/ip

mhaikelll committed:
Delete image due to name

mhaikelll committed:
Add back image for correct naming

mhaikelll committed:
Update README.md v4
